### ISSUE TITLE
MocktioExtension being too strict with stubs in `@BeforeEach` method

### ISSUE THREAD

**Issue by SamBarker** [mockito/mockito#1540]
I've written a test case for a class which persists errors to a file and thus I have a handful of tests to prove that non-errors cases are not persisted to a file however the majority of the tests in the test case are related to triggering various error scenarios. 

As the majority of test cases (~70%) require mock to supply the persistence file I've included a stub for that in the setup method (annotated using `junit5` `@BeforeEach`). I'm running the tests using `MockitoExtension` which fails tests with `org.mockito.exceptions.misusing.UnnecessaryStubbingException` for the 30% of test which demonstrate that the file isn't use when messaging works. 

While it is true that the stub is unused in those test I think it is discouraging readability by duplicating the stubbing to the majority of the tests in the class. From the [UnnecessaryStubbingException Javadoc](https://static.javadoc.io/org.mockito/mockito-core/2.23.4/org/mockito/exceptions/misusing/UnnecessaryStubbingException.html) 

> Mockito JUnit Runner triggers UnnecessaryStubbingException only when none of the test methods use the stubbings. This means that it is ok to put default stubbing in a 'setup' method or in test class constructor. That default stubbing needs to be used at least once by one of the test methods.

 I expected the Extension to offer the same behaviour.

I've contrived a simple example below which fails. 

I think this is related to: https://github.com/mockito/mockito/issues/769 but I think it is more of an issues with the `MockitoExtension` than strictness it self.

Dependencies on java 8 (8.0.171-oracle): 
```
<dependency>
            <groupId>org.junit.jupiter</groupId>
            <artifactId>junit-jupiter-api</artifactId>
            <version>5.3.1</version>
            <scope>test</scope>
        </dependency>
        <dependency>
            <groupId>org.junit.jupiter</groupId>
            <artifactId>junit-jupiter-engine</artifactId>
            <version>5.3.1</version>
            <scope>test</scope>
        </dependency><dependency>
        <groupId>org.mockito</groupId>
        <artifactId>mockito-core</artifactId>
        <version>2.23.4</version>
        <scope>test</scope>
    </dependency>
        <dependency>
            <groupId>org.mockito</groupId>
            <artifactId>mockito-junit-jupiter</artifactId>
            <version>2.23.4</version>
            <scope>test</scope>
        </dependency>
```

```
package nz.samandems.mockito;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.mockito.junit.jupiter.MockitoExtension;

@ExtendWith(MockitoExtension.class)
class MockitoLenienceyTestCase {

    @Mock
    private Dependency dependency;

    private SystemUnderTest systemUnderTest;


    @BeforeEach
    void setUp() {
        Mockito.when(dependency.doWork()).thenReturn(true);
        systemUnderTest = new SystemUnderTest(dependency);
    }


    @Test
    void shouldDoWork() {
        //Given

        //When
        systemUnderTest.doThing(true);

        //Then
        Mockito.verify(dependency).doWork();
    }

    @Test
    void shouldDoNoWork() {
        //Given

        //When
        systemUnderTest.doThing(false);

        //Then
        Mockito.verifyZeroInteractions(dependency);
    }


    private class SystemUnderTest {

        private final Dependency dependency;


        private SystemUnderTest(final Dependency dependency) {
            this.dependency = dependency;
        }


        public boolean doThing(boolean works) {
            if (works) {
                return dependency.doWork();
            }
            else {
                return false;
            }
        }
    }

    private interface Dependency {

        boolean doWork();
    }
}
```

**Comment by mockitoguy** [comment]
Thank you for reporting! I've read the JUnit5 Jupiter docs and it seems that before-class / after-class events are supported ([reference](https://junit.org/junit5/docs/current/user-guide/#extensions-lifecycle-callbacks)). Do you want to take a stab at it?

**Comment by SamBarker** [comment]
I'll try and make some time and take a look  😃

**Comment by SamBarker** [comment]
I've been digging around the mockito source for a bit tonight and have a failing test 😀 however I'm not sure the hint about before-class / after-class is actually relevant (I could well be missing quite a lot). 

I think the actual issue boils down to the fact that the extension uses the `UniversalTestListener`  which detects *all* unused stubs regardless [of location](https://github.com/mockito/mockito/blob/d77765088bdc747f25b30cf34fa9cbb4544ae35a/src/main/java/org/mockito/internal/junit/UnusedStubbingsFinder.java#L31-L41)
However the `StrictRunner` uses `UnnecessaryStubbingsReporter` which detects stubs [by location ](https://github.com/mockito/mockito/blob/d77765088bdc747f25b30cf34fa9cbb4544ae35a/src/main/java/org/mockito/internal/junit/UnusedStubbingsFinder.java#L53-L79).

I think its the different treatment of stubs by location which causes my contrived example to fail but I'm not at all sure what trade offs are involved between the two different ways of finding unused stubs.

**Comment by mockitoguy** [comment]
Thanks for looking into it! Yes, runner works differently than MockitoSession / JUnit rule / extension. They should reuse most code, though ;) Will you be able to change MockitoExtension implementation to do something like the runner? Once again, thank you very much for digging into it.

**Comment by SamBarker** [comment]
Ok I've spent a bit more time looking today and have established the Junit 4 Rule and the Jupiter extension operate the same way and have the same behaviour. I've also proved to my self that just changing the `UniversalTestListener` to detect unused stubbings by location isn't enough.

However I'm struggling to work out how to actually fix this. As far as I can tell the junit runner doesn't interact with the `MockitoSession` and observes a different life cycle. The runner appears to distinguish between the end a test "method" and the end of a test run/suite (I'm not sure what the proper terminology is here). 
> //only report when:
            //1. if all tests from given test have ran (filter requested is false)
            //   Otherwise we would report unnecessary stubs even if the user runs just single test from the class

See [StrictRunner](https://github.com/mockito/mockito/blob/7bb6cbe1f6f3e03409571981518e7312b625cc88/src/main/java/org/mockito/internal/runners/StrictRunner.java#L46)

As I understand things the a `MockitoSession` is only valid within the context of a single test "method" rather than a suite. As strict stubbing is validated in `finishMocking` I don't see a way for it to tell if the stub was valid for another test method even if it wasn't used in the current one.

Any pointers on what to look at / try would be greatly appreciated.

**Comment by mockitoguy** [comment]
>However I'm struggling to work out how to actually fix this. As far as I can tell the junit runner doesn't interact with the MockitoSession and observes a different life cycle. The runner appears to distinguish between the end a test "method" and the end of a test run/suite (I'm not sure what the proper terminology is here).

Correct, the runner does not use the session API. When I wrote the session API, JUnit rules where the gold standard and they don't have the beforeClass/afterClass events. Hence, the session API does not support it.

I'd say that we have 2 options:
a) change jupiter extension, stop using session API, start using the same stuff as the runner
b) upgrade session API somehow to support both cases: method level only (ala rule) and class level + method level (ala runner / jupiter)

Can you dig a bit deeper into doing a)? Your help is greatly appreciated!!! It will be a really nice feature for Jupiter integration.

**Comment by SamBarker** [comment]
I think I've ended up with a bit of a hybrid approach of `a` and `b` in https://github.com/mockito/mockito/pull/1574 

Looking at the changes again I can probably get a bit closer to `a` if desired but not sure how useful that is at this point.

**Comment by SamBarker** [comment]
I just wanted to check in to see if there were any thoughts on the strategy in https://github.com/mockito/mockito/pull/1574 as I realise the changes appeared in the holiday season.

**Comment by SamBarker** [comment]
I've opened a fresh pull request to address this issue: https://github.com/mockito/mockito/pull/1622 this is much more like (A) mentioned in this https://github.com/mockito/mockito/issues/1540#issuecomment-442523008 above

**Comment by SamBarker** [comment]
Ran out of git foo trying get #1622 updated to changes from release/2.x so have opened https://github.com/mockito/mockito/pull/1629 which is clean. 

Again this is a review of the impact of changes. I'm not very happy with the fact that it can no longer reporting stubbing errors at the test level, only at the session level.

**Comment by SamBarker** [comment]
Does anyone have a view as to wether #1629 is worth taking further? I ask as I have a couple of days available at work to take this forward if its appropriate. 

(That and the fact I've spent a bunch of time sprinkling `lenient = true` around tests I've been refactoring today)

**Comment by mockitoguy** [comment]
> That and the fact I've spent a bunch of time sprinkling lenient = true around tests I've been refactoring today

@SamBarker, what exact API you use to "sprinkle" lenient = true? Thx!

**Comment by SamBarker** [comment]
` @Mock(lenient = true)` on a field

Was what I was referring too.

**Comment by mockitoguy** [comment]
> @Mock(lenient = true) on a field

alternative:

```java
@BeforeEach
public void setup() {
   lenient().when(mock.foo()).thenReturn(true);
}
```

This way, only the specific stubbing is "lenient".

**Comment by SamBarker** [comment]
I like the sound of that it’s been a while since I looked at the case I was having issues with. I’ll try and use this version soon

**Comment by grimreaper** [comment]
ref #1924 which goes into this and other related issues

**Comment by TehBakker** [comment]
Any news on this ? 
Same probleme 90% of my tests need a when call and the other 10% throw exception.

I do not want to do lenient though as i feel UnnecessaryStubbingException is good just need to check with the whole tests

**Comment by SamBarker** [comment]
I never got a version of #1629 working in a way I was happy with. Looking at the details of #1924 I can see was trying to fix it in too small a scope. 

I do still think the original point stands but I think this bug should probably be closed in favour of #1924 .

**Comment by jingjiecb** [comment]
Any update on this issue?

**Comment by klemmons-gk** [comment]
Agreed; needs to be fixed. Migrating from junit4 to junit5 made every one of our test classes fail because of this issue. UnnecessaryStubbingException is being thrown for each test that doesn't use **everything** in setup() rather than throwing it once for each stub that is **never** used by any test.
